packages/optimizer/lib/transformers/AutoExtensionImporter.js

```diff
@@ -0,0 +1,269 @@
+            .forEach((a) => {
+              let attributeMapping = tagToAttributeMapping.get(tagName);
+              if (!attributeMapping) {
+                attributeMapping = [];
+                tagToAttributeMapping.set(tagName, attributeMapping);
+              }
+              attributeMapping.push(a);
+            });
+
+        // Map tags to bindable attributes which are named link `[text]`
+        const bindableAttributes = tag.attrs.filter((a) => a.name.startsWith('['))
+            .map((a) => a.name.substring(1, a.name.length - 1));
+        tagToBindAttributeMapping.set(tagName, new Set(bindableAttributes));
+      }
+      return {
+        extensionsMap,
+        tagToExtensionsMapping,
+        tagToAttributeMapping,
+        tagToBindAttributeMapping,
+      };
+    });
+  }
+
+  async transform(tree, params) {
+    if (!this.enabled) {
+      return;
+    }
+    const html = tree.root.firstChildByTag('html');
+    const head = html.firstChildByTag('head');
+    if (!head) return;
+
+    // Extensions which need to be imported
+    const extensionsToImport = new Set();
+    // Keep track of existing extensions imports to avoid duplicates
+    const existingImports = new Set();
+
+    // Some AMP components need to be detected in the head (e.g. amp-access)
+    await this.findExtensionsToImportInHead_(head, extensionsToImport, existingImports);
+
+    // Most AMP components can be detected in the body
+    const body = html.firstChildByTag('body');
+    await this.findExtensionsToImportInBody_(body, extensionsToImport);
+
+    if (extensionsToImport.length === 0) {
+      // Nothing to do
+      return;
+    }
+
+    // We use this for adding new import elements to the header
+    const referenceNode = findMetaViewport(head);
+
+    // Support custom runtime URLs
+    const host = calculateHost(params);
+    for (const extensionName of extensionsToImport) {
+      if (!existingImports.has(extensionName)) {
+        const extension = (await this.extensionSpec).extensionsMap.get(extensionName.trim());
+        this.log_.debug('auto importing', extensionName);
+        // Use the latest version by default
+        const version = extension.version[extension.version.length - 1];
+        const extensionImportAttribs = {
+          'async': '',
+          'src': `${host.ampUrlPrefix}/v0/${extensionName}-${version}.js`,
+        };
+        extensionImportAttribs[extension.type] = extensionName;
+        const extensionImport = tree.createElement('script', extensionImportAttribs);
+        head.insertAfter(extensionImport, referenceNode);
+      }
+    }
+  }
+
+  findExtensionsToImportInHead_(head, extensionsToImport, existingImports) {
+    let node = head.firstChild;
+    while (node) {
+      // Detect any existing extension imports
+      const customElement = this.getCustomElement_(node);
+      if (customElement) {
+        existingImports.add(customElement);
```

Comment:
Split existing import detection into a separate method.

Reply:
Hm, I think the overhead is practically non-existent (looping a list of ~20 items twice). But if you want to keep it this way, rename the method to something like `parseHead_` since it's not (just) finding extensions to import (unlike its body counterpart). The current name confused me when I was reading this code.

Reply:
renamed the method